### PR TITLE
Properly calculate alwaysUse24HourFormat on MacOS

### PR DESCRIPTION
**Pull request by kpsroka**
Moves the implementation if isAlwaysUse24HourFormat from iOS's FlutterViewController internals to common utility, and makes use of it on MacOS in order to return correct value of `alwaysUse24HourFormat`.

This PR partially resolves [#32006](https://github.com/flutter/flutter/issues/32006).

Note that on iOS 16+ and MacOS 13+, there is a new API for obtaining this information: https://developer.apple.com/documentation/foundation/locale/components/3952289-hourcycle. However, to keep things simpler, I wanted to not include changes to the logic.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by knopp — APPROVED** [review]
LGTM. This seems like a straightforward refactor. Not sure if this needs a test since it only moves things from iOS to darwin embedder. Also I'm not sure how to test this given that the behavior depends on system settings. @cbracken, any ideas?

**Comment by cbracken** [comment]
> Also I'm not sure how to test this given that the behavior depends on system settings.

In the usual case, we'd say extract out a thin utility class that wraps the system setting and then write a test of the usage site against a mock. In this case, this is just a refactoring that factors out the thin wrapper that wraps the system setting.

I think this is a fair case for requesting a test exemption.

**Review by cbracken — APPROVED** [review]
LGTM stamp from a Japanese personal seal

Looks great! Thanks for the patch!

File: shell/platform/darwin/common/framework/Headers/FlutterHourFormat.h [hunk 1]
```diff
@@ -0,0 +1,15 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_SHELL_PLATFORM_DARWIN_COMMON_FRAMEWORK_HEADERS_FLUTTERHOURFORMAT_H_
+#define FLUTTER_SHELL_PLATFORM_DARWIN_COMMON_FRAMEWORK_HEADERS_FLUTTERHOURFORMAT_H_
+
+#import <Foundation/Foundation.h>
+
+@interface FlutterHourFormat : NSObject
```
Comment (stuartmorgan) [inline shell/platform/darwin/common/framework/Headers/FlutterHourFormat.h:10]: Why is this a class? Classes that have nothing but static members is generally an anti-pattern, and unlike Java, Objective-C has stand-alone functions.

**Comment by stuartmorgan** [comment]
> In this case, this is just a refactoring that factors out the thin wrapper that wraps the system setting.

This isn't just a refactoring though; the purpose of this PR is to change the behavior of the macOS embedding. That's a behavioral change that should not regress, and thus should be tested.

**Comment by zanderso** [comment]
From PR triage: It looks like this PR is waiting for a test to be added, correct?

**Comment by kpsroka** [comment]
> From PR triage: It looks like this PR is waiting for a test to be added, correct?

Yes, I will get to that over the weekend.

File: shell/platform/darwin/common/framework/Headers/FlutterHourFormat.h (hunk 1 above)
Comment (cbracken) [reply shell/platform/darwin/common/framework/Headers/FlutterHourFormat.h:10]: Discussed over chat but I think it's reasonable for this to be a class in order to ease testing (injecting a fake in the mac test).

File: shell/platform/darwin/ios/framework/Source/FlutterViewControllerTest.mm [hunk 2]
```diff
@@ -1335,6 +1337,30 @@ - (void)testItReportsHighContrastWhenTraitCollectionRequestsIt {
   [mockTraitCollection stopMocking];
 }
 
+- (void)testItReportsAlwaysUsed24HourClock {
+  // Setup test.
+  id settingsChannel = OCMClassMock([FlutterBasicMessageChannel class]);
+  OCMStub([self.mockEngine settingsChannel]).andReturn(settingsChannel);
+
+  id mockHourFormat = OCMClassMock([FlutterHourFormat class]);
+  OCMStub([mockHourFormat isAlwaysUse24HourFormat]).andReturn(YES);
```
Comment (cbracken) [inline shell/platform/darwin/ios/framework/Source/FlutterViewControllerTest.mm:1346]: For thoroughness, consider testing both the `YES` and no case.

File: shell/platform/darwin/ios/framework/Source/FlutterViewControllerTest.mm [hunk 3]
```diff
@@ -170,6 +171,7 @@ - (void)triggerTouchRateCorrectionIfNeeded:(NSSet*)touches;
 @interface FlutterViewControllerTest : XCTestCase
 @property(nonatomic, strong) id mockEngine;
 @property(nonatomic, strong) id mockTextInputPlugin;
+@property(nonatomic, strong) id mockHourCycle;
```
Comment (cbracken) [inline shell/platform/darwin/ios/framework/Source/FlutterViewControllerTest.mm:174]: For clarity, consider naming this `mockHourFormat`.

File: shell/platform/darwin/ios/framework/Source/FlutterViewControllerTest.mm [hunk 4]
```diff
@@ -1335,6 +1337,30 @@ - (void)testItReportsHighContrastWhenTraitCollectionRequestsIt {
...
+                                                                     bundle:nil];
+  // Exercise behavior under test.
+  [vc traitCollectionDidChange:nil];
+
+  // Verify behavior.
+  OCMVerify([settingsChannel sendMessage:[OCMArg checkWithBlock:^BOOL(id message) {
+                               return message[@"alwaysUse24HourFormat"];
+                             }]]);
+
+  // Clean up mocks
```
Comment (cbracken) [inline shell/platform/darwin/ios/framework/Source/FlutterViewControllerTest.mm:1359]:
```suggestion
  // Clean up mocks.
```

This appears to be an issue with existing tests above, but we can fix for this patch at least. :)

File: shell/platform/darwin/macos/framework/Source/FlutterEngineTest.mm [hunk 5]
```diff
@@ -1268,6 +1268,32 @@ @implementation MockableFlutterEngine
...
+  OCMStub([channelMock sendMessage:[OCMArg any]]).andDo((^(NSInvocation* invocation) {
+    __weak id message;
+    [invocation getArgument:&message atIndex:2];
+    EXPECT_TRUE(message[@"alwaysUse24HourFormat"]);
+  }));
+
+  id engineMock = CreateMockFlutterEngine(nil);
+  EXPECT_TRUE([engineMock runWithEntrypoint:@"main"]);
+
+  // Clean up
```
Comment (cbracken) [inline shell/platform/darwin/macos/framework/Source/FlutterEngineTest.mm:1290]:
```suggestion
  // Clean up mocks.
```

**Review by cbracken — APPROVED** [review]
Overall looks good, just a few nits and suggestions.

File: shell/platform/darwin/ios/framework/Source/FlutterViewControllerTest.mm (hunk 3 above)
Comment (kpsroka) [reply shell/platform/darwin/ios/framework/Source/FlutterViewControllerTest.mm]: This is actually not used/needed/

**Comment by kpsroka** [comment]
I think that the PR is ready for merge. @cbracken please merge/add "autosubmit" [if you agree].

**Comment by kpsroka** [comment]
Note that https://github.com/flutter/flutter/issues/32006 is best left [re-]opened, and updated to refer only to `textScaleFactor`

**Review by cbracken — APPROVED** [review]

**Comment by cbracken** [comment]
Thanks for the fix, @kpsroka! Merged!

**Pull request merged**